pkger/service_models.go

```diff
@@ -0,0 +1,408 @@
+	}
+	sort.Slice(sum.Buckets, func(i, j int) bool {
+		return sum.Buckets[i].PkgName < sum.Buckets[j].PkgName
+	})
+
+	for _, l := range s.labels() {
+		if l.shouldRemove {
+			continue
+		}
+		sum.Labels = append(sum.Labels, l.summarize())
+	}
+	sort.Slice(sum.Labels, func(i, j int) bool {
+		return sum.Labels[i].PkgName < sum.Labels[j].PkgName
+	})
+
+	for _, m := range s.labelMappings {
+		sum.LabelMappings = append(sum.LabelMappings, m.summarize())
+	}
+	sort.Slice(sum.LabelMappings, func(i, j int) bool {
+		n, m := sum.LabelMappings[i], sum.LabelMappings[j]
+		if n.ResourceType != m.ResourceType {
+			return n.ResourceType < m.ResourceType
+		}
+		if n.ResourcePkgName != m.ResourcePkgName {
+			return n.ResourcePkgName < m.ResourcePkgName
+		}
+		return n.LabelName < m.LabelName
+	})
+
+	return sum
+}
+
+func (s *stateCoordinator) getLabelByPkgName(pkgName string) *stateLabel {
+	return s.mLabels[pkgName]
+}
+
+func (s *stateCoordinator) Contains(k Kind, pkgName string) bool {
+	_, ok := s.getObjectIDSetter(k, pkgName)
+	return ok
+}
+
+// setObjectID sets the id for the resource graphed from the object the key identifies.
+func (s *stateCoordinator) setObjectID(k Kind, pkgName string, id influxdb.ID) {
+	idSetFn, ok := s.getObjectIDSetter(k, pkgName)
+	if !ok {
+		return
+	}
+	idSetFn(id)
+}
+
+// setObjectID sets the id for the resource graphed from the object the key identifies.
+// The pkgName and kind are used as the unique identifier, when calling this it will
+// overwrite any existing value if one exists. If desired, check for the value by using
+// the Contains method.
+func (s *stateCoordinator) addObjectForRemoval(k Kind, pkgName string, id influxdb.ID) {
+	newIdentity := identity{
+		name: &references{val: pkgName},
+	}
+
+	switch k {
+	case KindBucket:
+		s.mBuckets[pkgName] = &stateBucket{
+			id:           id,
+			bucket:       &bucket{identity: newIdentity},
+			shouldRemove: true,
+		}
+	case KindLabel:
+		s.mLabels[pkgName] = &stateLabel{
+			id:           id,
+			label:        &label{identity: newIdentity},
+			shouldRemove: true,
+		}
+	}
+}
+
+func (s *stateCoordinator) getObjectIDSetter(k Kind, pkgName string) (func(influxdb.ID), bool) {
+	switch k {
+	case KindBucket:
+		r, ok := s.mBuckets[pkgName]
+		return func(id influxdb.ID) {
+			r.id = id
+		}, ok
+	case KindLabel:
+		r, ok := s.mLabels[pkgName]
+		return func(id influxdb.ID) {
+			r.id = id
+		}, ok
+	default:
+		return nil, false
+	}
+}
+
+type stateIdentity struct {
+	id           influxdb.ID
+	name         string
+	pkgName      string
+	resourceType influxdb.ResourceType
+	shouldRemove bool
+}
+
+func (s stateIdentity) exists() bool {
+	return s.id != 0
+}
+
+type stateBucket struct {
+	id, orgID    influxdb.ID
+	shouldRemove bool
+
+	existing *influxdb.Bucket
+
+	*bucket
+}
+
+func (b *stateBucket) diffBucket() DiffBucket {
+	diff := DiffBucket{
+		DiffIdentifier: DiffIdentifier{
+			ID:      SafeID(b.ID()),
+			Remove:  b.shouldRemove,
```

Comment:
not in this PR, but a follow up one, I want to change these Remove fields to be status, and have it identify how the resource impacts the platform. Felt like to much to include in this already large PR